File: src/main/scala/com/flink/tutorials/scala/chapter5_time/TriggerExample.scala
package com.flink.tutorials.scala.api.chapter5

import com.flink.tutorials.scala.api.chapter5.AggregateFunctionExample.AverageAggregate
import com.flink.tutorials.scala.utils.stock.{StockPrice, StockSource}
import org.apache.flink.api.common.state.{ValueState, ValueStateDescriptor}
import org.apache.flink.streaming.api.TimeCharacteristic
import org.apache.flink.streaming.api.scala._
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.windowing.triggers.{Trigger, TriggerResult}
import org.apache.flink.streaming.api.windowing.windows.TimeWindow

object TriggerExample {

  def main(args: Array[String]): Unit = {

    val senv = StreamExecutionEnvironment.getExecutionEnvironment
    senv.setStreamTimeCharacteristic(TimeCharacteristic.ProcessingTime)

    val stockStream = senv.addSource(new StockSource("stock/stock-tick-20200108.csv"))

    val average = stockStream
      .keyBy(s => s.symbol)
      .timeWindow(Time.seconds(60))
      .trigger(new MyTrigger)
      .aggregate(new AverageAggregate)

    average.print()

    senv.execute("trigger")
  }

  class MyTrigger extends Trigger[StockPrice, TimeWindow] {

    override def onElement(element: StockPrice,
                           time: Long,
                           window: TimeWindow,
                           triggerContext: Trigger.TriggerContext): TriggerResult = {
      val lastPriceState: ValueState[Double] = triggerContext.getPartitionedState(new ValueStateDescriptor[Double]("lastPriceState", classOf[Double]))

      // 设置返回默认值为CONTINUE
      var triggerResult: TriggerResult = TriggerResult.CONTINUE

      // 第一次使用lastPriceState时状态是空的,需要先进行判断
      // 状态数据由Java端生成，如果是空，返回一个null
      // 如果直接使用Scala的Double，需要使用下面的方法判断是否为空
      if (Option(lastPriceState.value()).isDefined) {
        if ((lastPriceState.value() - element.price) > lastPriceState.value() * 0.05) {
          // 如果价格跌幅大于5%，直接FIRE_AND_PURGE
          triggerResult = TriggerResult.FIRE_AND_PURGE
        } else if ((lastPriceState.value() - element.price) > lastPriceState.value() * 0.01) {
          val t = triggerContext.getCurrentProcessingTime + (10 * 1000 - (triggerContext.getCurrentProcessingTime % 10 * 1000))
          // 跌幅不大，注册一个10秒后的Timer
          triggerContext.registerProcessingTimeTimer(t)
        }
      }
      lastPriceState.update(element.price)
      triggerResult
    }

    // 这里我们不用EventTime，直接返回一个CONTINUE
    override def onEventTime(time: Long, window: TimeWindow, triggerContext: Trigger.TriggerContext): TriggerResult = {
      TriggerResult.CONTINUE
    }

    override def onProcessingTime(time: Long, window: TimeWindow, triggerContext: Trigger.TriggerContext): TriggerResult = {
      TriggerResult.FIRE_AND_PURGE
    }

    override def clear(window: TimeWindow, triggerContext: Trigger.TriggerContext): Unit = {
      val lastPrice: ValueState[Double] = triggerContext.getPartitionedState(new ValueStateDescriptor[Double]("lastPriceState", classOf[Double]))
      lastPrice.clear()
    }
  }
}

File: src/main/java/com/flink/tutorials/java/chapter5_time/TriggerExample.java
package com.flink.tutorials.java.chapter5_time;

import com.flink.tutorials.java.utils.stock.StockPrice;
import com.flink.tutorials.java.utils.stock.StockReaderFormat;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.api.common.state.ValueState;
import org.apache.flink.api.common.state.ValueStateDescriptor;
import org.apache.flink.api.common.typeinfo.Types;
import org.apache.flink.api.java.tuple.Tuple2;
import org.apache.flink.connector.file.src.FileSource;
import org.apache.flink.core.fs.Path;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.streaming.api.windowing.assigners.TumblingProcessingTimeWindows;
import org.apache.flink.streaming.api.windowing.triggers.Trigger;
import org.apache.flink.streaming.api.windowing.triggers.TriggerResult;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;

import java.time.Duration;

public class TriggerExample {

    public static void main(String[] args) throws Exception {

        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();

        String filePath = ClassLoader.getSystemResource("stock/stock-tick-20200108.csv")
                .getPath();
        FileSource<StockPrice> source = FileSource
                .forRecordStreamFormat(new StockReaderFormat(), new Path(filePath))
                .build();
        DataStream<StockPrice> stockStream = env
                .fromSource(source, WatermarkStrategy
                                .<StockPrice>forBoundedOutOfOrderness(Duration.ofSeconds(5)) // 允许5s乱序
                                .withTimestampAssigner((event, timestamp) -> event.ts)
                        , "StockSource");

        // 打印输入数据
        stockStream.print("Stock Data");

        DataStream<Tuple2<String, Double>> average = stockStream
                .keyBy(s -> s.symbol)
                .window(TumblingProcessingTimeWindows.of(Duration.ofSeconds(60)))
                .trigger(new MyTrigger())
                .aggregate(new AggregateFunctionExample.AverageAggregate());

        average.print();

        env.execute("trigger");
    }

    public static class MyTrigger extends Trigger<StockPrice, TimeWindow> {
        @Override
        public TriggerResult onElement(StockPrice element,
                                       long time,
                                       TimeWindow window,
                                       TriggerContext triggerContext) throws Exception {
            ValueState<Double> lastPriceState = triggerContext.getPartitionedState(
                    new ValueStateDescriptor<>("lastPriceState", Types.DOUBLE)
            );

            // 设置返回默认值为CONTINUE
            TriggerResult triggerResult = TriggerResult.CONTINUE;

            // 第一次使用lastPriceState时状态是空的,需要先进行判断
            // 如果是空，返回一个null
            if (null != lastPriceState.value()) {
                double priceChange = lastPriceState.value() - element.price;
                double dropPercent = priceChange / lastPriceState.value();

                System.out.printf(
                        "[Trigger] Symbol: %s | Current Price: %.2f | Last Price: %.2f | Drop: %.2f%%%n",
                        element.symbol, element.price, lastPriceState.value(), dropPercent * 100
                );

                if (dropPercent > 0.05) {
                    // 如果价格跌幅大于5%，直接FIRE_AND_PURGE
                    System.out.println("[Trigger] 5% drop triggered!");
                    triggerResult = TriggerResult.FIRE_AND_PURGE;
                } else if (dropPercent > 0.01) {
                    // 跌幅不大，注册一个10秒后的Timer
                    long t = triggerContext.getCurrentProcessingTime() + 10 * 1000;
                    System.out.printf("[Trigger] 1%% drop, register timer @ %d%n", t);
                    triggerContext.registerProcessingTimeTimer(t);
                }
            } else {
                System.out.printf(
                        "[Trigger] First element for symbol: %s | Price: %.2f%n",
                        element.symbol, element.price
                );
            }
            lastPriceState.update(element.price);
            return triggerResult;
        }

        // 这里我们不用EventTime，直接返回一个CONTINUE
        @Override
        public TriggerResult onEventTime(long time, TimeWindow window, TriggerContext triggerContext) {
            return TriggerResult.CONTINUE;
        }

        @Override
        public TriggerResult onProcessingTime(long time, TimeWindow window, TriggerContext triggerContext) {
            System.out.printf("[Timer] Triggered at: %d%n", time);
            return TriggerResult.FIRE_AND_PURGE;
        }

        @Override
        public void clear(TimeWindow window, TriggerContext triggerContext) {
            ValueState<Double> lastPriceState = triggerContext.getPartitionedState(
                    new ValueStateDescriptor<>("lastPriceState", Types.DOUBLE)
            );
            lastPriceState.clear();
        }
    }
}